cluster_tools/tests/test_all.py

```diff
@@ -166,174 +201,151 @@ def maybe_negate(b: bool) -> bool:
 
 
 def get_pid() -> int:
     return os.getpid()
 
 
-def test_process_id() -> None:
+def test_process_id(exc_with_pickling: cluster_tools.Executor) -> None:
+    exc = exc_with_pickling
     outer_pid = os.getpid()
 
-    def compare_pids(executor: cluster_tools.Executor) -> None:
-        with executor:
-            future = executor.submit(get_pid)
-            inner_pid = future.result()
-
-            should_differ = not isinstance(exc, cluster_tools.DebugSequentialExecutor)
+    with exc:
+        future = exc.submit(get_pid)
+        inner_pid = future.result()
 
-            if should_differ:
-                assert (
-                    inner_pid != outer_pid
-                ), f"Inner and outer pid should differ, but both are {inner_pid}."
-            else:
-                assert (
-                    inner_pid == outer_pid
-                ), f"Inner and outer pid should be equal, but {inner_pid} != {outer_pid}."
+        should_differ = not isinstance(
+            exc,
+            (
+                cluster_tools.SequentialExecutor,
+                cluster_tools.SequentialPickleExecutor,
+            ),
+        )
 
-    for exc in get_executors(with_debug_sequential=True):
-        compare_pids(exc)
+        if should_differ:
+            assert (
+                inner_pid != outer_pid
+            ), f"Inner and outer pid should differ, but both are {inner_pid}."
+        else:
+            assert (
+                inner_pid == outer_pid
+            ), f"Inner and outer pid should be equal, but {inner_pid} != {outer_pid}."
+
+
+def test_unordered_sleep(exc: cluster_tools.Executor) -> None:
+    with exc:
+        durations = [5, 1]
+        futures = [exc.submit(sleep, n) for n in durations]
+        results = [f.result() for f in exc.as_completed(futures)]
+
+        if isinstance(
+            exc,
+            (
+                cluster_tools.SequentialExecutor,
+                cluster_tools.SequentialPickleExecutor,
+            ),
+        ):
+            # futures.as_completed does not return previously completed futures in completion order.
+            # For sequential executors as_completed is only called after all futures completed, though.
+            results.sort()
+
+        durations.sort()
+        for duration, result in zip(durations, results):
+            assert result == duration
+
+
+def test_map_to_futures(exc: cluster_tools.Executor) -> None:
+    with exc:
+        durations = [5, 1]
+        futures = exc.map_to_futures(sleep, durations)
+        results = [f.result() for f in exc.as_completed(futures)]
 
+        if isinstance(
+            exc,
+            (cluster_tools.SequentialExecutor, cluster_tools.SequentialPickleExecutor),
+        ):
+            # futures.as_completed does not return previously completed futures in completion order.
+            # For sequential executors as_completed is only called after all futures completed, though.
+            results.sort()
```

Comment:
thanks for your patience to test this again! i looked into the implementation of `as_completed` and found that it returns the results in reverse order if the futures are already finished. see [here](https://github.com/python/cpython/blob/f85af035c5cb9a981f5e3164425f27cf73231b5f/Lib/concurrent/futures/_base.py#L181).

my suggestion would be: split the test into two. the first test only uses async executors (i.e., not the sequential ones). the second test basically tests for the synchronous submit as the unordered sleep stuff doesn't make much sense for that. e.g., like this:

```python
future = exc.submit(sleep, some_duration)
assert future.done()
assert future.result() == some_duration
```

what do you think?

Reply:
Thanks for looking that up. It's certainly an interesting design decision :upside_down_face: 

Splitting up the test makes sense :+1:

Reply:
I ended up not splitting the test, because I wanted to avoid additional complexity for the executor fixtures. However, I reverted the sorting, made sure to only use `as_completed` for the async executors and replaced the assertion for loop. I hope that's fine.